File: backend/app/api/v1/deployments.py
from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends, status

from app.api.deps import get_deployment_service
from app.schemas.deployments import (
    DeploymentPromoteRequest,
    DeploymentPromotionRead,
    DeploymentRollbackRead,
    DeploymentRollbackRequest,
)
from app.services.deployments import DeploymentService

router = APIRouter(prefix="/deployments", tags=["deployments"])

DeploymentDependency = Annotated[DeploymentService, Depends(get_deployment_service)]


@router.post(
    "/promote", response_model=DeploymentPromotionRead, status_code=status.HTTP_201_CREATED
)
def promote_deployment(
    payload: DeploymentPromoteRequest,
    service: DeploymentDependency,
) -> DeploymentPromotionRead:
    return service.promote_version(
        agent_id=payload.agent_id,
        agent_version_id=payload.agent_version_id,
        reason=payload.reason,
    )


@router.post("/rollback", response_model=DeploymentRollbackRead)
def rollback_deployment(
    payload: DeploymentRollbackRequest,
    service: DeploymentDependency,
) -> DeploymentRollbackRead:
    return service.rollback(
        agent_id=payload.agent_id,
        reason=payload.reason,
    )

File: frontend/lib/api/deployments.ts
import { requestJson } from "@/lib/api/client";
import type {
  DeploymentEventRead,
  DeploymentPromotionRead,
  DeploymentRollbackRead,
} from "@/lib/api/types";

export async function promoteDeployment(payload: {
  agent_id: string;
  agent_version_id: string;
  reason?: string | null;
}): Promise<DeploymentPromotionRead> {
  return requestJson<DeploymentPromotionRead>("/api/v1/deployments/promote", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function rollbackDeployment(payload: {
  agent_id: string;
  reason?: string | null;
}): Promise<DeploymentRollbackRead> {
  return requestJson<DeploymentRollbackRead>("/api/v1/deployments/rollback", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function listAgentDeployments(agentId: string): Promise<DeploymentEventRead[]> {
  return requestJson<DeploymentEventRead[]>(`/api/v1/agents/${agentId}/deployments`);
}

